feat: show databases CLI/API

hiltontj opened the pull request:
_Follows #25737 (keeping in draft until that merges)_

Closes #25745 

This PR provides both a CLI and underlying API for listing databases in the InfluxDB 3 Core server. Details are below.

There was already a method to list databases for the query executor for InfluxQL; this works by exposing that via the `HttpApi` in `influxdb3_server`.

However, one thing that we may address is that the query result for that uses `iox::database` as the column name. If we are removing references to `iox`, then we may want to just have it as `database`. I left it as is, for now, because I wanted to keep code churn down and wasn't sure why we use that prefix in the first place for the `SHOW DATABASES` and `SHOW RETENTION POLICIES` InfluxQL queries.

## Details

### CLI

This PR provides the `influxdb3 show` CLI:
```
influxdb3 show -h
List resources on the InfluxDB 3 Core server

Usage: influxdb3 show <COMMAND>

Commands:
  databases  List databases
  help       Print this message or the help of the given subcommand(s)

Options:
  -h, --help  Print help information
```
with the ability to list databases:
```
influxdb3 show databases -h
List databases

Usage: influxdb3 show databases [OPTIONS]

Options:
  -H, --host <HOST_URL>         The host URL of the running InfluxDB 3 Core server [env: INFLUXDB3_HOST_URL=] [default: http://127.0.0.1:8181]
      --token <AUTH_TOKEN>      The token for authentication with the InfluxDB 3 Core server [env: INFLUXDB3_AUTH_TOKEN=]
      --show-deleted            Include databases that were marked as deleted in the output
      --format <OUTPUT_FORMAT>  The format in which to output the list of databases [default: pretty] [possible values: pretty, json, json_lines, csv]
  -h, --help                    Print help information
```
Since this uses the query executor, we can pass a `--format` argument to get the output as JSON, CSV, or JSONL, but by default, it uses the `pretty` format:
```
influxdb3 show databases
+---------------+
| iox::database |
+---------------+
| bar           |
+---------------+
```
The `--show-deleted` flag will have the `deleted` column displayed as well as any databases that have been marked as deleted:
```
influxdb3 show databases --show-deleted
+---------------------+---------+
| iox::database       | deleted |
+---------------------+---------+
| bar                 | false   |
| foo-20250105T202949 | true    |
+---------------------+---------+
```

### API

The API to list databases can be invoked via:
```
GET /api/v3/configure/database
```
with optional parameters:
* `format`: `pretty`, `json`, `csv`, `parquet`, or `jsonl`
* `show_deleted`: `bool`, defaults to `false`

Note that `database` is singular in the API endpoint, to be consistent with the other database related create/delete API endpoints. We could change it to be plural `databases` if that is the convention we want to go with.

pauldix commented on influxdb3/src/commands/show.rs:
```diff
@@ -0,0 +1,82 @@
+
+#[derive(Debug, Parser)]
+pub enum SubCommand {
+    /// View all active databases
```
Looking further down the PR, it looks like active means all databases not marked as deleted. Should the language be more specific to say that? If so, same change for other spots where "active" is used.

hiltontj replied:
Agreed. I addressed in https://github.com/influxdata/influxdb/pull/25748/commits/0d722a259466bd6706d020d9552022813a59ef51

However, shouldn't we also list out the DBs that have been deleted? We could include a `deleted` column in the display, instead of just the `iox::database` column.

This would be useful for having visibility on what has been deleted.

hiltontj replied:
> However, shouldn't we also list out the DBs that have been deleted? We could include a deleted column in the display, instead of just the iox::database column.

FWIW, we could do that on a separate issue/PR. I think that being able to list DBs is somewhat needed as several people have asked about it and having to use InfluxQL is a bit odd - so would be good to get this in.

pauldix reviewed (approved):
Good stuff, just one small comment

pauldix commented:
I think including deleted DBs with the flag in the output would be good

On Sun, Jan 5, 2025 at 9:30 AM Trevor Hilton ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In influxdb3/src/commands/show.rs
> <https://github.com/influxdata/influxdb/pull/25748#discussion_r1903282493>
> :
>
> > @@ -0,0 +1,82 @@
> +use clap::{Parser, ValueEnum};
> +use secrecy::{ExposeSecret, Secret};
> +use std::error::Error;
> +use url::Url;
> +
> +#[derive(Debug, Parser)]
> +pub struct Config {
> +    #[clap(subcommand)]
> +    cmd: SubCommand,
> +}
> +
> +#[derive(Debug, Parser)]
> +pub enum SubCommand {
> +    /// View all active databases
>
> Agreed. I addressed in 0d722a2
> <https://github.com/influxdata/influxdb/commit/0d722a259466bd6706d020d9552022813a59ef51>
>
> However, shouldn't we also list out the DBs that have been deleted? We
> could include a deleted column in the display, instead of just the
> iox::database column.
>
> This would be useful for having visibility on what has been deleted.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/influxdata/influxdb/pull/25748#discussion_r1903282493>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAABB242XGRVS4ZYNTVKBID2JE6YBAVCNFSM6AAAAABUTOTPFGVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDKMZQHEYTAMBTGM>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

hiltontj commented:
> I think including deleted DBs with the flag in the output would be good

I went ahead and did that in https://github.com/influxdata/influxdb/pull/25748/commits/2ffd596d78cbecea4da2e4610bd33c2e7cc589da, updated PR description with sample output.

peterbarnett03 commented:
If we're going to incorporate deletion status, the column approach is odd to me, as that should be false essentially forever; once true, the row-entry is removed in its entirety eventually. 

I'd suggest a temporary bottom table portion for clarity, which works as a better indicator of soft-deletion and  recoverability, plus less things to think about for the primary use case: showing table names. 


```
+-------------------------------------+
| Active Databases                    |
+-------------------------------------+
| bar                                 | 
| foo                                 |
+-------------------------------------+
| Databases Pending Deletion          |
+-------------------------------------+
| foo-20250104T11583918               |
+-------------------------------------+
```

pauldix commented:
It comes through as a single table, so the formatting is what does that. It would be strange to have special formatting for this. I would say that the results should be sorted by `deleted, name` so that non-deleted databases show up first in the listing.

peterbarnett03 commented:
Ok. Not worth major revision work regardless. Having deleted sorted first (false, true) rather than alphabetical name works well for the same visual approach.

hiltontj commented:
Another idea I had was to have a --show-deleted flag in the CLI. By default the deleted ones do not show up in the result (nor does the deleted column) unless that flag is passed in the CLI command.

This could be handled with a projection on the generated query result.

pauldix commented:
Oh, the flag makes sense to me

hiltontj commented:
Added the `--show-deleted` flag in https://github.com/influxdata/influxdb/pull/25748/commits/5945428c0029796d8f0cb522f3c1394af3c2c3c0.

Now, by default, it only shows the non-deleted databases in the output and will not display the `deleted` column, unless that flag is passed. Sorting is first by deleted (false, then true) then by name.

I updated the PR description with CLI outputs.

hiltontj commented:
Thanks for all the feedback @pauldix and @peterbarnett03 - I will get this merged following #25737
